tools/approval-service/internal/eventprocessor/workflowevents.go

```diff
@@ -0,0 +1,433 @@
+		return nil, fmt.Errorf("storing workflow info: %w", err)
+	}
+
+	reason, err := approvalHandler.generateAccessRequestReason(e.WorkflowID, e.Environment)
+	if err != nil {
+		return nil, fmt.Errorf("generating access request reason: %w", err)
+	}
+	newReq.SetRequestReason(reason)
+
+	created, err := w.teleClient.CreateAccessRequestV2(ctx, newReq)
+	if err != nil {
+		return nil, fmt.Errorf("creating access request: %w", err)
+	}
+
+	return created, nil
+}
+
+// tryStartEventProcessing attempts to start processing an event if it's not already being processed.
+// Returns true if processing should proceed, false if the event is already being processed.
+// This method provides deduplication to prevent concurrent processing of the same workflow event.
+func (w *WorkflowEventsProcessor) tryStartEventProcessing(eventID string) bool {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if _, exists := w.currentlyProcessing[eventID]; exists {
+		// Already processing this event, skip it.
+		return false
+	}
+
+	// Mark this event as currently being processed.
+	w.currentlyProcessing[eventID] = struct{}{}
+	return true
+}
+
+// finishEventProcessing marks an event as finished processing, allowing it to be processed again in the future.
+// This removes the event from the currently processing set to prevent memory leaks.
+func (w *WorkflowEventsProcessor) finishEventProcessing(eventID string) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	delete(w.currentlyProcessing, eventID)
+}
+
+// onAccessRequestReviewed processes the Access Request review event.
+func (w *WorkflowEventsProcessor) onAccessRequestReviewed(ctx context.Context, req types.AccessRequest) {
+	info, err := w.store.GetWorkflowInfo(ctx, req)
+	if err != nil {
+		// If we cannot find the workflow info, we cannot process the request.
+		// This is likely due to the access request not having the required labels.
+		w.log.Debug("Getting workflow info for access request", "access_request_name", req.GetName(), "error", err)
+		return
+	}
+
+	approvalHandler, err := w.getApprovalHandler(info.Org, info.Repo)
+	if err != nil {
+		w.log.Error("Error getting approval handler for access request", "access_request_name", req.GetName(), "org", info.Org, "repo", info.Repo, "error", err)
+		return
+	}
+
+	if err := approvalHandler.handleDecisionForAccessRequestReviewed(ctx, req.GetState(), info.Env, info.WorkflowRunID); err != nil {
+		w.log.Error("Error handling access request reviewed", "access_request_name", req.GetName(), "error", err)
+		return
+	}
+	w.log.Info("Handled access request reviewed", "access_request_name", req.GetName(), "org", info.Org, "repo", info.Repo)
+}
+
+func (w *WorkflowEventsProcessor) getApprovalHandler(org, repo string) (*GitHubDeploymentApprovalHandler, error) {
+	handler, ok := w.deploymentApprovalHandlers[formatGitHubRepoKey(org, repo)]
+	if !ok {
+		return nil, fmt.Errorf("the organization %q and repository %q are not configured for deployment approval", org, repo)
+	}
+	return handler, nil
+}
```

Comment:
Like I suggested above, this should go away. Let's just make it work for teleport and worry about others later.